client/websocket.go

```diff
@@ -0,0 +1,147 @@
+type WebsocketGateway struct {
+	reader chan []byte
+	ctx    context.Context
+
+	conn *websocket.Conn
+	// There can only be one writer on the connection at a time.
+	// In case the outer context gets cancelled there can be a situation
+	// in which writing the contexts error cause can collide
+	// with a normal write to the websocket.
+	writeLock sync.Mutex
+}
+
+// NewWebsocketGateway returns a new websocket wrapper allowing to easily write and consume
+// messages to/from the underlying websocket connection.
+// It allows providing a context which is cancelled as soon as the underlying websocket connection
+// is closed by either side of the connection.
+func NewWebsocketGateway(ctx context.Context, conn *websocket.Conn) *WebsocketGateway {
+	gw := &WebsocketGateway{
+		reader: make(chan []byte),
+		conn:   conn,
+	}
+
+	gwCtx, gwCancel := context.WithCancelCause(ctx)
+	gw.ctx = gwCtx
+
+	go func() {
+		<-gwCtx.Done()
+
+		// Send close control message.
+		// Try to send the cause from the outer context if present.
+		msg := websocket.FormatCloseMessage(websocket.CloseInternalServerErr, context.Cause(gwCtx).Error())
+
+		gw.writeLock.Lock()
+		_ = gw.conn.WriteMessage(websocket.CloseMessage, msg)
+		gw.writeLock.Unlock()
+
+		// Shutdown the read loop.
+		_ = gw.conn.Close()
+	}()
+
+	go func() {
+		defer close(gw.reader)
+
+		for {
+			_, reader, err := conn.ReadMessage()
+			if err != nil {
+				// If the connection got closed due to the outer context, return this error instead.
+				if ctx.Err() != nil {
+					// Try to use the cause from the outer context if present.
+					err = context.Cause(ctx)
+				}
+
+				// If the other end sent a close control message try to extract the error.
+				// Such control messages are sent base64 encoded to allow sending special
+				// characters like quotes.
+				closeErr, ok := err.(*websocket.CloseError)
+				if ok {
+					errorStr, decodeErr := base64.StdEncoding.DecodeString(closeErr.Text)
+					if decodeErr == nil {
+						err = fmt.Errorf("%s", errorStr)
+					}
+				}
+
+				// Cancel the inner context too with the respective error.
+				defer gwCancel(err)
+				return
+			}
+
+			gw.reader <- reader
+		}
+	}()
+
+	return gw
+}
+
+// Receive returns the inner channel which allows reading from the websocket connection.
+// If used together with other channels ensure to also consume the gateway's context
+// in order to get informed about a potentially closed connection.
+// If there aren't other channels that need to be consumed in parallel use ReceiveCombined instead.
+func (w *WebsocketGateway) Receive() <-chan []byte {
+	return w.reader
+}
+
+// ReceiveCombined tries to read from the websocket connection and unmarshals
+// the received data into v.
+// It's waiting on both the websocket connection and the context and returns
+// whatever is returning/cancelled first.
+func (w *WebsocketGateway) ReceiveCombined(v any) error {
```

Comment:
Happy to change the name but I don't think another context should be passed here.

The overall problem comes from the fact that the underlying websocket `*websocket.Conn` doesn't know the concept of a context (remember [this](https://github.com/canonical/microcluster/pull/237#discussion_r1722949800) comment).
That was the main reason for adding the `WebsocketGateway` to be able to couple an active websocket connection with an outer context.

Therefore when you create the `WebsocketGateway` you pass the outer context which will be cancelled in case we observe any errors reading from the websocket connection. This is the intended approach according to [the docs](https://pkg.go.dev/github.com/gorilla/websocket#Conn.NextReader):

> Applications must break out of the application's read loop when this method returns a non-nil error value. Errors returned from this method are permanent. Once this method returns a non-nil error, all subsequent calls to this method return the same error.

To be able to react on errors from the underlying websocket quickly, a go routine is spawned that continuously reads from it and cancels the context in case of error.

So you can reuse the context which was already passed when creating the `WebsocketGateway`.
In case you want to timeout waiting for a message (or error of websocket connection), you can use this pattern:

```go
gw := NewWebsocketGateway(outerCtx, conn)

select {
    case <-gw.Context().Done():
       // outer context cancelled or error in websocket conn
    case <-gw.Receive():
       // actual message received
    case <-time.After(duration):
       // timeout waiting for message
}
``` 

But one use case I would see is wrapping the timeout concept in `ReceiveWithContext(ctx, v)` so you can do something like "I want to wait 10s for a message".
So within the func we will wait on both the `gw.Context().Done()` and the passed `ctx` as well as the `w.Receive()` chan.

Reply:
I have now implemented `ReceiveWithContext` which allows having an additional context for timeout scenarios.

But for this PR it's not used as we solely rely on the outer context from the gateway itself.

Reply:
So I think part of my confusion here is that we had `ReceiveCombined` and `Receive`, but the context is internal, so why do we have the distinction between the two functions?`ReceiveWithContext` is also a victim of this distinction because it's now using the same context for two different `select` cases in all its usages.

So I'm wondering why we even need to call `Receive` apart from `ReceiveWithContext(gw.Context()...` from anywhere outside the `client` package? 

Could we just move the internal context handling directly into `Receive`? Then we can do away with `ReceiveWithContext`, and if we ever need to include it, we can add it back and just have it `select` over the context and `gw.Receive`?

Reply:
The reason for `Receive` as a standalone is to get a channel from the websocket that allows you to either wait on an ingress message from the websocket or on a different channel using a `select` statement.

This is required in the initiating session.
We can at the same time receive:
 * a new join intent from any joiner which has to be relayed to the client to be displayed in the table
 * the message from the client confirming the join intents

See the section of the code [here](https://github.com/canonical/microcloud/pull/383/files#diff-27dcd3e94dd8f08cf82109cf63ca764783c103a170caa9c30738b1ba87f5c6f8R101).

Now all `ReceiveWithContext` usages use the same context, but I found it might be useful in the future to supply an additional context with a timeout if you want to wait for a message to arrive in X seconds.